File: online/LeetCodeCN/1~100/70.php
class Solution {
    private $dp;

    public function __construct() {
        $this->dp = [1 => 1, 2 => 2];
    }

    /**
     * @param Integer $n
     * @return Integer
     */
    public function climbStairs($n) {
        $ans = $this->dp[$n];

        if ($ans == 0) {
            $ans = $this->dp[$n] = $this->climbStairs($n-1) + $this->climbStairs($n-2);
        }

        return $ans;
    }
}

File: online/LeetCodeCN/1~100/70.go
// 爬楼梯 https://leetcode-cn.com/problems/climbing-stairs/
// DP

var dp = map[int]int {
    1: 1,
    2: 2,
}

func climbStairs(n int) int {
    ans, cached := dp[n]
    if !cached {
        ans = climbStairs(n-1) + climbStairs(n-2)
        dp[n] = ans
    }
    return ans
}
